### ISSUE TITLE
The state of service map is awful

### ISSUE THREAD

**Issue by gracicot** [gracicot/kangaru#28]
Currently, kangaru suggest users to define the whole service maps. This has the effect of the user needing to "send" the service map to the container and can easily lead to fragmented maps.

Another problem is that the service map won't follow basic parameter rules like cv qualifier. This needs to be fixed, especially in the version 4.

**Comment by gracicot** [comment]
We'll call that feature _unified service map_

**Comment by gracicot** [comment]
We could use ADL to get the return type of a function that receive the service and return the definition.

It could look like this:

```c++
auto service_map(MyClass&) -> MyClassService;
```

**Comment by gracicot** [comment]
I thought about it and introducing a map like that may drop a useful feature that users might have use before: multiple maps.

You see, saying that there is one function that takes a parameter and returns the service prevent using multiple maps.

To fix that problem, we may add an optional parameter to the map; instead of the function only receiving the parameter, we could optionally receive a tag. Something like this:

```c++
auto service_map(MyClass&) -> MyClassService; // default
auto service_map(MyClass&, kgr::Map<AlternativeMap>) -> MyClassService2; // mapping for AlternativeMap
```

**Comment by gracicot** [comment]
I wonder if there's a way to integrate it in the current major version. It would simplify declaration of services so much!

**Comment by gracicot** [comment]
I'll explore the idea to si if it's possible to implement without breaking code.

**Comment by gracicot** [comment]
It's possible to implement, but one cannot define specialized map (aka alternative service maps). Specialized service maps would be too hard to implement as current template parameters are all asking for a template template parameter (argh why I did this??)

**Comment by gracicot** [comment]
Fixed in dev-3.2.x. Note that multiple maps are not supported yet, because it would require breaking.
